### ISSUE TITLE
File/Folder owner not restored properly

### ISSUE THREAD

**Issue by robertfoss** [restic/restic#2447]
Output of `restic version`
--------------------------
```
$ restic version
restic 0.9.5 compiled with go1.12.9 on linux/amd64
```

How did you run restic exactly?
-------------------------------
```
REPOSITORY=/mnt/server/restic
BACKUP_CONTENT="\
	/home/$USER/Apps \
	/etc/udev/rules.d/49-*.rules \
	/opt \
	"
sudo restic -r $REPOSITORY \
	backup \
	--verbose \
	--tag `hostname` \
	--tag `date +%Y-%m-%d` \
	$BACKUP_CONTENT
sudo restic -r $REPOSITORY restore latest
```
What backend/server/service did you use to store the repository?
----------------------------------------------------------------
Local file, the local file in question happens to be a SMB-share.

Expected behavior
-----------------
I expect the restore to replicate the file ownership as it was created.

Actual behavior
---------------
There are two issues here:

### Backup has to be done as root
Needing to backup any files that belong to another user forces you to run `restic backup` as root.
This in turn will force you to run `restic restore` as root too since you need root permissions to read the restic repository now. This in turn means that even if you want to fix some of the issues restic has with permissions by splitting the restore by original owner will be problematic since the `restic restore` _always_ has run as root.
```
sudo restic -r $REPOSITORY restore $ID --target / --include /home/ --include /opt/
sudo restic -r $REPOSITORY restore $ID --target / --exclude /home/ --exclude /opt/
```
Even when manually intervening as in the above example, the files that used to belong to $USER are now going to be restored as being owner by the user root, since `restic restore` has to be run as root  due to repository permissions.

### Original file owner is not restored
After a `restic backup` -> OS reinstall -> `restic restore` none of the original file ownership attributes are restored.

I haven't inspected the restic codebase, but it seems to me that no ownership data is stored for the backup up content. If it is stored, it isn't used it during a restore.

My backups contain things like builds of linux OSes, and these require the owners of files to be correct, if they are changed the restored files will not be usable any longer due to ownership issues.

Do you have any idea what may have caused this?
-----------------------------------------------
Owner of  is not maintained properly over a backup+restore cycle.

Do you have an idea how to solve the issue?
-------------------------------------------
### Backup has to be done as root
I would suggest the following solution: `restic backup` really does need to be run as root to read the files. But there is no reason for the restic repository files to be created in such a way that no other user can read them. Possibly the permission could be set in a loose way like `chmod 777` on folders and `chmod 555` files in the restic repository.

### Original file owner is not restored
As a first step I would like to see the restored files have the same UID & GID as the backed up file. 

#### Mapping Posix Owner Usernames to UIDs
As a second step I would like to see is the username/UID & group/GID mapping getting translated between backup and restore, in order to support a restore happening on a freshly installed OS where UIDs and GIDs may be different from the original installation.
```
restic backup: robertfoss (UID 1001)
restic backup: root (UID 0)
restic restore: robertfoss (UID 1002)
restic restore: root (UID 0)
```
In the above case I would to see all files backed up using UID 1001, to be restored as UID 1002.
And of course all files backed up as UID 0, to be restored as UID 0.

Did restic help you or made you happy in any way?
-------------------------------------------------
Apart from that I don't really have any complaints, restic is a pretty solid backup solution where no good alternatives used to exist.

**Comment by cdhowie** [comment]
Permissions and ownership is restored, but as a second step after restore completes.  If the restore is aborted or fails for whatever reason, permissions and ownership will not have been restored.

Did the restore get aborted, fail, or were you inspecting the permissions/ownership while the restore was still running?

**Comment by robertfoss** [comment]
Ah! Well, than all is sort of well and good.
The restore is sort of a 20h process, so I inspected it during the restore since I wasn't aware of this behavior.

**Comment by robertfoss** [comment]
@cdhowie It would be better if permissions and ownership was restored simultaneously however.

**Comment by bernd-wechner** [comment]
I found this for same reason. Was inspecting the restore result while it was running and confused.  It is an interesting design decision to do this as a second pass of all the restored files and not as the files are restored one by one. Perhaps there was a good reason? Even if so, it would be nice to have an option then to restore only the permission and owners (i.e. do pass 2 only) in case pass 1 is interrupted say and you'd like to at least get the partial restore properly configured.

I had another interesting issue in that I backed up /data and then restored to /data and ended up with /data/data ;-). Easy fixed with a mv when done, but mildly surprising.
